### PR TITLE
VIMC-4266: separate out impact calculations into individual functions

### PR DESCRIPTION
**Pull request by r-ash**

### CODE REVIEW

File: R/impact_central.R [hunk 1]
```diff
@@ -152,22 +152,16 @@ impact_by_year_of_vaccination <- function(meta1, raw_impact, fvps, fvps_updates
     fvps_updates <- fvps_updates[fvps_updates$year %in% vaccination_years, ]
   }
 
-
+  raw_impact$birth_cohort <- raw_impact$time
```
Comment (r-ash) [inline R/impact_central.R:155]: In the new functions I'm using `birth_cohort` as the column name instead of `time` which gives this awkward renaming step.

I see Xiang has it hardcoded to check that `time` column exists above https://github.com/vimc/vimpact/pull/7/files#diff-94b59a1086fde6063424f28c1f56d3bbR113 I'm happy to drop this change and just use `time` everywhere if that is a term that is familiar to science as representing `birth_cohort` otherwise I think it is better to be more specific in the naming

File: R/impact_central.R [hunk 2]
```diff
@@ -152,22 +152,16 @@ impact_by_year_of_vaccination <- function(meta1, raw_impact, fvps, fvps_updates
...
-    cohort_impact <- raw_impact[raw_impact$time %in% (min(fvps$time):max(fvps$time)), c("country", "time", "burden_outcome", "value")]
-    cohort_fvps <- stats::aggregate(fvps ~ country + time, fvps, sum, na.rm = TRUE)
-    d <- merge_by_common_cols(cohort_impact, cohort_fvps, all = TRUE)
+    d <- impact_by_year_of_vaccination_cohort_perspective(raw_impact, fvps,
+                                                          vaccination_years)
   }
-  d$impact_ratio <- d$value / d$fvps
+  d$time <- d$birth_cohort
+  d$birth_cohort <- NULL
   d$value <- NULL
```
Comment (r-ash) [inline R/impact_central.R:165]: Once you've had a chance to check this looks okay I will remove `value` and `fvps` from return of the `imapct_by_year*` functions but leaving it in now as the seeing which numbers are coming through makes it a bit easier to see what those functions are doing

File: R/impact_central.R [hunk 3]
```diff
@@ -202,3 +196,106 @@ determine_vaccine_delivery <- function(meta1){
 #
 #
 # }
+
+#' Calculate impact by year of vaccination country perspective
+#'
+#' This will calculate the impact by year of vaccination by country and
+#' burden outcome for a single disease and vaccine.
```
Comment (r-ash) [inline R/impact_central.R:203]: Is it useful to write up the method here?

File: tests/testthat/test-impact.R [hunk 4]
```diff
@@ -123,4 +123,141 @@ test_that("test if vimpact functions are working as expected for central estimat
...
+  expect_equal(unique(impact$country), c("ETH", "PAK"))
+  expect_equal(unique(impact$burden_outcome), "deaths")
+  expect_equal(impact$value, c(1614, 3508))
+  expect_equal(impact$fvps, c(199, 417))
+  expect_equal(impact$impact_ratio, c(8.1, 8.4), tolerance = 1.e-1)
+
+  ## Impact for routine filters on birth cohorts in range of
+  ## year of vaccinations - min age fvps
+  fvps$age <- rep(1, 10)
+  impact <- impact_by_year_of_vaccination_country_perspective(
```
Comment (r-ash) [inline tests/testthat/test-impact.R:143]: I'm curious about this edge case as in calculating the impact ratio the impact is summed from 2 values, from the 2004 and 2005 data but the fvps used are only the 2005 ones - which gives a much higher impact_ratio than the calculation above - but maybe that is intended?

File: tests/testthat/test-impact.R [hunk 5]
```diff
@@ -123,4 +123,141 @@ test_that("test if vimpact functions are working as expected for central estimat
...
+  expect_equal(impact$impact_ratio,
+               c(6.9, 8.4, 10.1, 4.3, 15.0, 8.6, 4.6, 5.4, 5.5, 2.8),
+               tolerance = 1e-1)
+
+  ## Impact calculation filters on impact calculations within range of
+  ## birth cohort
+  raw_impact <- raw_impact[raw_impact$year == 2001, ]
+  impact <- impact_by_year_of_vaccination_cohort_perspective(raw_impact, fvps,
+                                                             2000:2030)
+  expect_equal(impact,
```
Comment (r-ash) [inline tests/testthat/test-impact.R:210]: Similarly this edge case is kind of odd - what do we really want to return here? I'm guessing we're not really interested in those rows where impact ratio is NA

**Review by r-ash — COMMENTED** [review]
Few notes it would be useful to get your feedback on. Particularly around edge cases where some of the output looks a bit odd to me.

Also would be helpful if you can run all the tests on your end as the existing tests get skipped on the CI system at the moment. In particular double checking the previously existing `test-impact.R` test works would be useful.

File: R/impact_central.R [hunk 6]
```diff
@@ -202,3 +196,106 @@ determine_vaccine_delivery <- function(meta1){
...
+#'
+#' @param raw_impact Data frame of raw impact data this needs to have
+#' columns country, value, burden_outcome and either year & age or birth_cohort
+#' @param fvps Data frame of fully vaccination person data with columns
+#' country, fvps and either year & age or birth_cohort
+#' @param activity_type `routine` or `campaign` activity type
+#' @param vaccination_years Years of vaccination of interest
+#'
+#' @return Impact ratio by country and burden outcome
+#' @export
```
Comment (r-ash) [inline R/impact_central.R:216]: Really it would be ideal to have some examples here - I wonder if we have any representative data we can pull from the DB and store in the `inst` dir? I can then use it here in example

File: R/impact_central.R (hunk 1 above)
Comment (kgaythorpe) [reply R/impact_central.R:155]: I like the switch to birth_cohort

File: R/impact_central.R (hunk 2 above)
Comment (kgaythorpe) [reply R/impact_central.R:165]: I would be happy to keep fvps in general- makes it easier to see what is going on even if it is unnecessary

File: R/impact_central.R (hunk 3 above)
Comment (kgaythorpe) [reply R/impact_central.R:203]: excitingly we reverted the name to "activity perspective" but I think having a sentence on the method may be helpful

File: R/impact_central.R (hunk 6 above)
Comment (xiangli313) [reply R/impact_central.R:216]: I am not sure how sensible it is to pull real data into the package. If we want, maybe we have to keep country names anonymous, and downscale numbers by a random factor?

**Review by xiangli313 — APPROVED** [review]
I am happy with Rob's updates. In terms of sample data, let me know if you would like me to try to include one - maybe only one anonymous country, a few diseases, a single touchstone.

**Comment by r-ash** [comment]
Thanks Xiang, let's think about how we can add test data next time we do some development on this we think it would be useful for.
